File: rock63/src/main/java/com/uksusoff/rock63/data/DataSource.java
package com.uksusoff.rock63.data;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;

import com.j256.ormlite.android.apptools.OpenHelperManager;
import com.j256.ormlite.stmt.DeleteBuilder;
import com.uksusoff.rock63.data.entities.EventItem;
import com.uksusoff.rock63.data.entities.NewsItem;
import com.uksusoff.rock63.data.entities.VenueItem;
import com.uksusoff.rock63.exceptions.NoContentException;
import com.uksusoff.rock63.exceptions.NoInternetException;
import com.uksusoff.rock63.utils.DateUtils;
import com.uksusoff.rock63.utils.StringUtils;

import org.androidannotations.annotations.AfterInject;
import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;
import org.androidannotations.annotations.sharedpreferences.Pref;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLConnection;
import java.sql.SQLException;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;

@EBean(scope = EBean.Scope.Singleton)
public class DataSource {

    private static final String BASE_API_URL = "https://rock63.ru/api";

    private static final long NEWS_LIFETIME_DAYS = 60;

    private DBHelper database;

    @RootContext
    Context context;

    @Pref
    InternalPrefs_ intPrefs;

    @AfterInject
    public void init() {
        database = OpenHelperManager.getHelper(context, DBHelper.class);
    }

    public List<NewsItem> getAllNews() {
        try {
            return database.getNewsItemsDao().queryBuilder().orderBy("date", false).query();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private void newsCleanUp() throws SQLException {
        Date before = new Date();

        before.setTime(before.getTime() - NEWS_LIFETIME_DAYS * 1000 * 60 * 60 * 24);

        DeleteBuilder<NewsItem, Integer> builder = database.getNewsItemsDao().deleteBuilder();

        builder.where().le("date", before);
        builder.delete();
    }

    private void newsRefresh() throws NoInternetException, NoContentException {
        final SQLiteDatabase db = database.getWritableDatabase();

        db.beginTransaction();

        try {
            JSONArray news = getEntitiesArray("/news");

            newsCleanUp();

            List<Integer> ids = new LinkedList<>();
            for (NewsItem newsItem : database.getNewsItemsDao().queryForAll()) {
                ids.add(newsItem.id);
            }

            for (int i = 0; i < news.length(); i++) {
                JSONObject newsItemJson = news.getJSONObject(i);

                int id = newsItemJson.getInt("id");

                NewsItem newsItem;
                if (ids.contains(id)) {
                    newsItem = database.getNewsItemsDao().queryForId(id);
                } else {
                    newsItem = new NewsItem();
                }

                newsItem.id = newsItemJson.getInt("id");
                newsItem.title = newsItemJson.getString("title");
                newsItem.url = newsItemJson.getString("url");
                newsItem.body = newsItemJson.has("desc") ? newsItemJson.getString("desc") : null;
                newsItem.ext = newsItemJson.has("ext_url") ? newsItemJson.getString("ext_url") : null;
                newsItem.date = DateUtils.fromTimestamp(newsItemJson.getInt("date_p"));
                newsItem.thumbnailSmall = newsItemJson.has("img") ? newsItemJson.getJSONObject("img").getString("img_s") : null;
                newsItem.thumbnailMiddle = newsItemJson.has("img") ? newsItemJson.getJSONObject("img").getString("img_m") : null;
                newsItem.isNew = true;

                database.getNewsItemsDao().createOrUpdate(newsItem);
            }

            db.setTransactionSuccessful();
        } catch (JSONException | SQLException e) {
            throw new RuntimeException(e);
        } finally {
            db.endTransaction();
        }
    }

    public List<EventItem> eventsGetAll(boolean ascending) {
        try {
            return database.getEventItemsDao().queryBuilder().orderBy("start", ascending).query();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private void eventsRefresh() throws NoInternetException, NoContentException {
        final SQLiteDatabase db = database.getWritableDatabase();

        db.beginTransaction();

        try {
            JSONArray events = getEntitiesArray("/events");

            database.getEventItemsDao().deleteBuilder().delete();

            for (int i = 0; i < events.length(); i++) {
                JSONObject eventJson = events.getJSONObject(i);

                if (eventJson.has("venues_up")) {
                    long lastPlacesUpdate = eventJson.getLong("venues_up");
                    if (intPrefs.lastUpdatedPlaces().get() != lastPlacesUpdate) {
                        this.venuesRefresh();
                        intPrefs.lastUpdatedPlaces().put(lastPlacesUpdate);
                    }
                }

                JSONObject date = eventJson.getJSONObject("date");

                EventItem eventItem = new EventItem();
                eventItem.id = eventJson.getInt("id");
                eventItem.title = eventJson.getString("title");
                eventItem.url = eventJson.getString("url");
                eventItem.body = eventJson.has("desc") ? eventJson.getString("desc") : null;
                eventItem.ext = eventJson.has("ext_url") ? eventJson.getString("ext_url") : null;
                eventItem.notify = eventJson.has("notify");
                eventItem.start = DateUtils.fromTimestamp(date.getInt("s"));
                eventItem.end = date.has("e") ? DateUtils.fromTimestamp(date.getInt("e")) : null;
                eventItem.thumbnailSmall = eventJson.has("img") ? eventJson.getJSONObject("img").getString("img_s") : null;
                eventItem.thumbnailMiddle = eventJson.has("img") ? eventJson.getJSONObject("img").getString("img_m") : null;
                if (eventJson.has("v_id")) {
                    eventItem.setVenueItem(database.getVenueItemsDao().queryForId(eventJson.getInt("v_id")));
                } else {
                    eventItem.setVenueItem(null);
                }

                database.getEventItemsDao().create(eventItem);
            }

            db.setTransactionSuccessful();
        } catch (JSONException | SQLException e) {
            throw new RuntimeException(e);
        } finally {
            db.endTransaction();
        }
    }

    public EventItem eventGetRelated(NewsItem newsItem) {
        try {
            return database.getEventItemsDao().queryForId(newsItem.id);
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private void venuesRefresh() throws NoInternetException, NoContentException {
        try {
            JSONArray venues = getEntitiesArray("/venues");

            database.getVenueItemsDao().deleteBuilder().delete();

            for (int i = 0; i < venues.length(); i++) {
                JSONObject venueJson = venues.getJSONObject(i);

                VenueItem venueItem = new VenueItem();
                venueItem.id = venueJson.getInt("id");
                venueItem.title = venueJson.getString("title");
                venueItem.address = venueJson.getString("address");
                if (venueJson.has("site") && !venueJson.getString("site").isEmpty()) {
                    venueItem.url = venueJson.getString("site");
                }
                if (venueJson.has("phone")) {
                    venueItem.phone = venueJson.getString("phone");
                }
                if (venueJson.has("vk") && !venueJson.getString("vk").isEmpty()) {
                    venueItem.vk = venueJson.getString("vk");
                }
                if (venueJson.has("latitude") && venueJson.has("longitude")) {
                    venueItem.latitude = venueJson.getString("latitude");
                    venueItem.longitude = venueJson.getString("longitude");
                }

                database.getVenueItemsDao().create(venueItem);
            }
        } catch (JSONException | SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private JSONArray getEntitiesArray(String endpoint) throws NoInternetException, NoContentException {
        String contents;

        try {
            URLConnection conn = new URL(BASE_API_URL + endpoint).openConnection();
            InputStream in = conn.getInputStream();
            contents = StringUtils.fromStream(in);
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new NoInternetException();
        }

        if (contents.isEmpty()) {
            throw new NoContentException();
        }

        try {
            return new JSONArray(contents);
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    public void sourcesRefresh() throws NoInternetException, NoContentException {
        this.venuesRefresh();
        this.eventsRefresh();
        this.newsRefresh();
    }
}

File: rock63/src/main/java/com/uksusoff/rock63/data/DataSource.kt
package com.uksusoff.rock63.data

import android.content.Context
import com.j256.ormlite.android.apptools.OpenHelperManager
import com.uksusoff.rock63.data.entities.Event
import com.uksusoff.rock63.data.entities.NewsItem
import com.uksusoff.rock63.data.entities.Place
import com.uksusoff.rock63.utils.CommonUtils
import org.androidannotations.annotations.AfterInject
import org.androidannotations.annotations.EBean
import org.androidannotations.annotations.RootContext
import org.androidannotations.annotations.sharedpreferences.Pref
import org.json.JSONArray
import org.json.JSONException
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL
import java.net.URLConnection
import java.sql.SQLException
import java.text.SimpleDateFormat
import java.util.*

@EBean(scope = EBean.Scope.Singleton)
open class DataSource {

    private lateinit var database: DBHelper

    @RootContext
    protected lateinit var context: Context
    @Pref
    protected lateinit var intPrefs: InternalPrefs_

    @AfterInject
    fun init() {
        database = OpenHelperManager.getHelper(context, DBHelper::class.java)
    }

    @get:Throws(SQLException::class)
    val allNews: List<NewsItem>
        get() = database.newsItemDao!!.queryBuilder().orderBy("date", false).query()

    @Throws(SQLException::class)
    fun clearOldNews() {
        val before = Date()
        before.time = before.time - NEWS_LIFETIME_DAYS * 1000 * 60 * 60 * 24
        val builder = database.newsItemDao!!.deleteBuilder()
        builder.where().le("date", before)
        builder.delete()
    }

    @Throws(NoInternetException::class)
    fun refreshNews() {
        val optstore = context.getSharedPreferences(ROCK63_OPTIONS_STORE, 0)
        val lastNewsUpdate = optstore.getInt(ROCK63_OPTION_LAST_NEWS_UPDATE, 0)
        val contents: String
        val conn: URLConnection
        try {
            val url: URL
            url = if (lastNewsUpdate != 0) {
                val d = CommonUtils.getDateFromTimestamp(lastNewsUpdate)
                URL(String.format(NEWS_SOURCE_URL_FROM_DATE, SimpleDateFormat("yyyyy/MM/dd", Locale.getDefault()).format(d)))
            } else {
                URL(NEWS_SOURCE_URL)
            }
            conn = url.openConnection()
            val `in` = conn.getInputStream()
            contents = CommonUtils.convertStreamToString(`in`)
        } catch (e: MalformedURLException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw NoInternetException()
        }
        if (!contents.isEmpty()) {
            val db = database.writableDatabase
            db.beginTransaction()
            try {
                val news = JSONArray(contents)
                clearOldNews()
                val ids: MutableList<Int> = LinkedList()
                for (item in database.newsItemDao!!.queryForAll()) {
                    ids.add(item.id)
                }
                for (i in 0 until news.length()) {
                    val newsItemJson = news.getJSONObject(i)
                    val id = newsItemJson.getInt("id")
                    var newsItem: NewsItem
                    newsItem = if (ids.contains(id)) {
                        database.newsItemDao!!.queryForId(id)
                    } else {
                        NewsItem()
                    }
                    newsItem.id = id
                    newsItem.date = CommonUtils.getDateFromTimestamp(newsItemJson.getInt("date_p"))
                    if (newsItemJson.has("img")) {
                        newsItem.smallThumbUrl = newsItemJson.getJSONObject("img").getString("img_s")
                        newsItem.mediumThumbUrl = newsItemJson.getJSONObject("img").getString("img_m")
                    }
                    newsItem.title = newsItemJson.getString("title")
                    var body: String? = null
                    if (newsItemJson.has("desc")) {
                        body = newsItemJson.getString("desc")
                    }
                    if (newsItemJson.has("ext_url")) {
                        body = if (body == null) "" else "$body "
                        body += newsItemJson.getString("ext_url")
                    }
                    newsItem.body = body
                    if (newsItemJson.has("url")) {
                        newsItem.url = newsItemJson.getString("url")
                    }
                    newsItem.isNew = true
                    database.newsItemDao!!.createOrUpdate(newsItem)
                }
                db.setTransactionSuccessful()
            } catch (e: JSONException) {
                throw RuntimeException(e)
            } catch (e: SQLException) {
                throw RuntimeException(e)
            } finally {
                db.endTransaction()
            }
            val editor = optstore.edit()
            editor.putInt(ROCK63_OPTION_LAST_NEWS_UPDATE, lastNewsUpdate)
            editor.apply()
        }
    }

    fun getAllEvents(ascending: Boolean): List<Event> {
        return try {
            database.eventDao!!.queryBuilder().orderBy("start", ascending).query()
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }
    }

    val allEvents: List<Event>
        get() = getAllEvents(true)

    @Throws(NoInternetException::class)
    fun refreshEvents() {
        val contents: String
        val conn: URLConnection
        try {
            conn = URL(EVENTS_SOURCE_URL).openConnection()
            val `in` = conn.getInputStream()
            contents = CommonUtils.convertStreamToString(`in`)
        } catch (e: MalformedURLException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw NoInternetException()
        }
        if (!contents.isEmpty()) {
            val db = database.writableDatabase
            db.beginTransaction()
            try {
                database.eventDao!!.deleteBuilder().delete()
                val events = JSONArray(contents)
                for (i in 0 until events.length()) {
                    val eventJson = events.getJSONObject(i)
                    if (eventJson.has("venues_up")) {
                        val lastPlacesUpdate = eventJson.getLong("venues_up")
                        if (intPrefs.lastUpdatedPlaces().get() != lastPlacesUpdate) {
                            refreshPlacesSync()
                            intPrefs.lastUpdatedPlaces().put(lastPlacesUpdate)
                        }
                    }
                    val e = Event()
                    e.id = eventJson.getInt("id")
                    e.title = eventJson.getString("title")
                    if (eventJson.has("desc")) {
                        e.body = eventJson.getString("desc")
                    } else {
                        e.body = ""
                    }
                    if (eventJson.has("ext_url")) {
                        e.body = e.body + eventJson.getString("ext_url")
                    }
                    e.start = CommonUtils.getDateFromTimestamp(eventJson.getJSONObject("date").getInt("s"))
                    if (eventJson.has("img")) {
                        e.mediumThumbUrl = eventJson.getJSONObject("img").getString("img_m")
                    }
                    if (eventJson.getJSONObject("date").has("e")) e.end = CommonUtils.getDateFromTimestamp(eventJson.getJSONObject("date").getInt("e"))
                    if (eventJson.has("v_id")) {
                        e.place = database.placeDao!!.queryForId(eventJson.getInt("v_id"))
                    } else {
                        e.place = null
                    }
                    e.url = eventJson.getString("url")
                    e.isNotify = eventJson.has("notify")
                    database.eventDao!!.create(e)
                }
                db.setTransactionSuccessful()
            } catch (e: JSONException) {
                throw RuntimeException(e)
            } catch (e: SQLException) {
                throw RuntimeException(e)
            } finally {
                db.endTransaction()
            }
        }
    }

    fun getRelatedEvent(item: NewsItem): Event {
        return try {
            database.eventDao!!.queryForId(item.id)
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }
    }

    @Throws(NoInternetException::class)
    fun refreshPlacesSync(): Boolean {
        val contents: String
        val conn: URLConnection
        try {
            conn = URL(PLACES_SOURCE_URL).openConnection()
            val `in` = conn.getInputStream()
            contents = CommonUtils.convertStreamToString(`in`)
        } catch (e: MalformedURLException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw NoInternetException()
        }
        return if (!contents.isEmpty()) {
            try {
                val places = JSONArray(contents)
                database.placeDao!!.deleteBuilder().delete()
                for (i in 0 until places.length()) {
                    val placeJson = places.getJSONObject(i)
                    val place = Place()
                    place.id = placeJson.getInt("id")
                    place.name = placeJson.getString("title")
                    place.address = placeJson.getString("address")
                    if (placeJson.has("site")) {
                        place.url = placeJson.getString("site")
                    }
                    if (placeJson.has("phone")) {
                        place.phone = placeJson.getString("phone")
                    }
                    if (placeJson.has("vk")) {
                        place.vkUrl = placeJson.getString("vk")
                    }
                    database.placeDao!!.create(place)
                }
            } catch (e: JSONException) {
                throw RuntimeException(e)
            } catch (e: SQLException) {
                throw RuntimeException(e)
            }
            true
        } else false
    }

    class NoInternetException : Exception()
    companion object {
        private const val TAG = "DataSource"
        private const val NEWS_SOURCE_URL = "http://rock63.ru/api/news"
        private const val NEWS_SOURCE_URL_FROM_DATE = "http://rock63.ru/api/news"
        private const val EVENTS_SOURCE_URL = "http://rock63.ru/api/events"
        private const val PLACES_SOURCE_URL = "http://rock63.ru/api/venues"
        const val NO_PLACE_ID = -1
        private const val NEWS_LIFETIME_DAYS: Long = 60
        const val ROCK63_OPTIONS_STORE = "rock63_options_store"
        const val ROCK63_OPTION_LAST_NEWS_UPDATE = "rock63_options_last_news_update"
    }
}